Charge FAQ

FilipeFcp opened the pull request:
Hi all. This is the Charge FAQ.
@marc-flex , this is the solver most distant from my expertise. I hope it doesn't need many corrections, but feel free to point out any changes. Also, feel free to request any other topics or questions you’d like covered.

@alec-flexcompute I think you are in Maine, right? Feel free to look at it when you are back in the office.

FilipeFcp committed:
Charge FAQ

marc-flex commented on _faqs/how-can-i-define-a-semiconductor-material.md:
```diff
@@ -0,0 +1,54 @@
+
+### Density of States and Band Gap Energy
+
+- `N_c` (PositiveFloat) – Effective density of states in the conduction band.  
+  *Units: cm⁻³*
+
+- `N_v` (PositiveFloat) – Effective density of states in the valence band.  
+  *Units: cm⁻³*
+
+- `E_g` (PositiveFloat) – Band gap energy.  
+  *Units: eV*
```
This is going to change soon when we introduce other (temperature-dependent) models. Will we need to update this by hand? Or can this somehow be linked to the docs?

FilipeFcp replied:
Probably by hand. @daquinteroflex , do you think there is a relatively easy way for me to implement some automation here?

daquinteroflex replied:
Would suggest to have a LLM prompt to convert the API automatically and do it manually when needed

marc-flex commented on _faqs/how-can-i-define-a-semiconductor-material.md:
```diff
@@ -0,0 +1,54 @@
+
+### Mobility Models
+
+Mobility can be temperature-dependent (based on the [Caughey-Thomas mobility model](None){: .color-primary-hover}, implemented with the class [`CaugheyThomasMobility`](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.CaugheyThomasMobility.html){: .color-primary-hover}) or constant (implemented with the class [`ConstantMobilityModel`](https://docs.flexcompute.com/proj ...
```
The Caughey-Thomas mobility is not only dependent on temperature but also on doping. We should add that it accounts for thermal and doping effects

marc-flex commented on _faqs/how-can-i-define-a-semiconductor-material.md:
```diff
@@ -0,0 +1,54 @@
+
+### Recombination Models
+
+Recombination mechanisms can include:
```
Maybe change to `Electron-hole recombination mechanisms` ?

marc-flex commented on _faqs/how-can-i-define-a-semiconductor-material.md:
```diff
@@ -0,0 +1,54 @@
+
+- `delta_E_g` – [Slotboom model for band-gap narrowing](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.SlotboomBandGapNarrowing.html){: .color-primary-hover}
+
+- `N_a` (PositiveFloat) – Number of acceptor charges (holes).  
+  *Units: cm⁻³*
+
+- `N_d` (PositiveFloat) – Number of donor charges (electrons).  
+  *Units: cm⁻³*
```
These can not only be floats. They can also be doping boxes: https://docs.flexcompute.com/projects/tidy3d/en/latest/api/index.html#doping

marc-flex commented on _faqs/how-can-i-define-a-semiconductor-material.md:
```diff
@@ -0,0 +1,54 @@
+- `N_d` (PositiveFloat) – Number of donor charges (electrons).  
+  *Units: cm⁻³*
+
+For a practical example, please refer to [this example notebook](https://www.flexcompute.com/tidy3d/examples/notebooks/ThermoOpticDopedModulator/){: .color-primary-hover}.
```
I would rather direct users to the ChargeSolver notebook?

marc-flex commented on _faqs/how-can-i-define-the-simulation-grid.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: How Can I Define the Simulation Grid?
+date: 2025-06-24 19:15:49
+enabled: true
+category: "Charge"
+---
+There are two classes available for defining the grid specification:
```
This looks good but maybe we need to move this somewhere else. Specially now that we're going to introduce a new pipeline. Plus we have the new mesh refinements:

- https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.GridRefinementRegion.html#tidy3d.GridRefinementRegion
- https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.GridRefinementLine.html#tidy3d.GridRefinementLine

These refinements should be added I think. You can refer to the updated ChargeSolver notebook. What do you think @momchil-flex

FilipeFcp replied:
I see. I will look at these new classes

marc-flex commented on _faqs/how-do-i-set-a-charge-simulation.md:
```diff
@@ -0,0 +1,34 @@
+- **Define meshing specifications**  
+  The simplest option is to use the [UniformUnstructuredGrid](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.UniformUnstructuredGrid.html){: .color-primary-hover} mesh type, where you only need to specify the grid size `dl`. The mesher will automatically generate a mesh that fits the structures while respecting the given resolution.
+
+- **Create the simulation object**  
+  Create a [HeatChargeSimulation](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.HeatChargeSimulation.html){: .color-primary-hover} object.
```
Before creating the simulation object I'd say that you also need to specify an analysis type of Charge type. Currently there's only IsothermalSteadyChargeSimulation

FilipeFcp replied:
Sure. Can you clarify what happens with the Default option for the `analysis_spec`  (`None`)? It will use a default [IsothermalSteadyChargeDCAnalysis](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.IsothermalSteadyChargeDCAnalysis.html#tidy3d.IsothermalSteadyChargeDCAnalysis)

marc-flex replied:
So when `analysis_spec` isn't specified (`None`) we look for medium definitions, boundary conditions, monitors, and sources to determine whether we should run a Conduction or Heat simulation. 

When you specify `analysis_spec` we do it the other way around. i.e., if it is specified to something (currently we can define Unsteady Heat simulations, isothermal Charge simulations, and soon non isothermal Charge) we check that mediums, BCs, monitors, sources and so on are consistent with the defined analysis type.

alec-flexcompute replied:
Should the details of analysis type be included in this doc? I think yes, but could also be its own FAQ

marc-flex commented on _faqs/what-types-of-charge-simulations-are-available.md:
```diff
@@ -0,0 +1,15 @@
+enabled: true
+category: "Charge"
+---
+Currently, there are two types of Charge simulations that can be carried out using the [HeatChargeSimulation](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.HeatChargeSimulation.html){: .color-primary-hover} class:
+
+- **Conduction**:  
+    Solves the electrical conduction equation.
+
+- **Charge**:  
+    For structures containing a [SemiconductorMedium](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.SemiconductorMedium.html){: .color-primary-hover}, the drift-diffusion equations are also solved.
```
I'm torn by this. I'd say that Conduction and Charge are two different types of simulations. Within Charge we have isothermal and (hopefully) soon, non isothermal. In the future there'll also be other types. So not sure if should include the Conduction in these types. Also, because Conduction jobs  don't require `analysis_spec`.

I get how this may be confusing. Maybe @dbochkov-flexcompute and @momchil-flex can give their opinion on this.

FilipeFcp replied:
I think by "two types of Charge simulations" I should say "two types of simulations can be carried out using the CHARGE solver", or something like that. 

And I might have misunderstood this part. What I understood is that the conduction equation is always solved, and for semiconductor media, the drift-diffusion equation is also solved.

marc-flex commented on _faqs/which-types-of-charge-boundary-conditions-are-available.md:
```diff
@@ -0,0 +1,47 @@
+{% endhighlight %}
+{% include copy-button.html %}</div>
+
+Note that the `voltage` argument can be a string, in which case all voltages will be simulated.
```
An array not a string, I think

marc-flex commented on _faqs/which-types-of-charge-boundary-conditions-are-available.md:
```diff
@@ -0,0 +1,47 @@
+{% endhighlight %}
+{% include copy-button.html %}</div>
+
+The `current` argument must be a `float`.
```
`current` currently doesn't work with Charge, only Conduction. With regards to the previous comment, I think it'd be best to focus on Charge and leave Conduction for a different faq?

marc-flex reviewed (commented):
Thanks @FilipeFcp for all this work!

I guess the main comment is that we're mixing here Conduction and Charge. I'd leave Conduction out but we can discuss this offline. 

A different comment may be that this will be hard to maintain going forward, right? I mean, since we're continuously developing this solver :S

marc-flex commented:
Basically, in Charge and Conduction we're solving sets of equations, even if the conduction equation is in practice very similar to one of the Charge equations. 

Apart from that. there are other internal differences, which is why I'd prefer to have Conduction and Charge separated.

alec-flexcompute reviewed (approved):
Looks good @FilipeFcp ! I'm also confused as to the charge/conduction simulations, as well as which future features will happen soon. However, aside from my comment about analysis specs, it looks good and will be very useful!

daquinteroflex commented:
@marc-flex this repo is public, but we should make it private really

FilipeFcp committed:
comments

FilipeFcp commented:
Hi @marc-flex 

I believe I’ve addressed all comments.

I removed the section on “what types of charge simulations are available”. I can work on the Conduction part separately.

Once the other classes then  [IsothermalSteadyChargeDCAnalysis](https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.IsothermalSteadyChargeDCAnalysis.html#tidy3d.IsothermalSteadyChargeDCAnalysis), are implemented, I will create an FAQ specifically for the analysis types.

Some parts were automated to make future updates easier, but the process isn’t fully automatic; I still need to run the script manually, but I think I can do it when we release a new version.

marc-flex commented:
Thanks @FilipeFcp!
